File: backend-go/test/state_test.go
package test

import (
"testing"
"time"

"backend-go/internal/state"
)

func TestEmergencyWins(t *testing.T) {
event := state.Event{Type: "panic", Source: "mobile", TS: time.Now()}
candidate := state.ApplyRules(event, nil, true)
if candidate.Status != "EMERGENCY" {
t.Fatalf("expected EMERGENCY, got %s", candidate.Status)
}
}

func TestPriorityOrder(t *testing.T) {
event := state.Event{Type: "assigned", Source: "task", TS: time.Now()}
current := &state.Snapshot{Status: "AVAILABLE", Source: "GEOFENCE", TS: time.Now().Add(-time.Minute)}
candidate := state.ApplyRules(event, current, true)
winner := state.ChooseByPriority(candidate, current, time.Now())
if winner.Status != "AVAILABLE" {
t.Fatalf("expected existing status to win due to higher priority")
}
}

File: backend-node/test/state.test.js
import { expect } from 'chai';
import { applyRules, chooseByPriority, PRIORITIES } from '../src/server.js';

describe('state engine', () => {
  it('prioritises emergency events', () => {
    const event = { type: 'panic', source: 'mobile' };
    const current = { status: 'ON_SHIFT', source: 'TASK', ts: new Date().toISOString() };
    const result = applyRules(event, current, true);
    expect(result.status).to.equal('EMERGENCY');
    expect(result.priority).to.equal('EMERGENCY');
  });

  it('chooses by priority order', () => {
    const candidate = { status: 'EMERGENCY', priority: 'EMERGENCY', reason: 'panic' };
    const current = { status: 'BUSY', source: 'TASK', ts: new Date(Date.now() - 1000).toISOString() };
    const winner = chooseByPriority(candidate, current, new Date().toISOString());
    expect(winner.status).to.equal('EMERGENCY');
  });

  it('exports priority order', () => {
    expect(PRIORITIES).to.include('EMERGENCY');
  });
});
